### PR TITLE
[CBRD-25164] Add new testcase for Error results from covered index queries using function indexes.

### PR DESCRIPTION
**Pull request by tw-kang**
http://jira.cubrid.org/browse/CBRD-25164

### CODE REVIEW

File: sql/_13_issues/_24_1h/cases/cbrd_25164.sql [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+insert into tx values (1,10,100), (2,20,200);
+insert into ty values (1,1,1000), (2,2,2000);
+create index idx on tx(ln(va));
+update statistics on tx;
+select ty.id, tx.va, ty.vc from tx join ty on tx.id = ty.tx_id where ln(tx.va) > 2;
+
+drop table tx;
+drop table ty;
+
+-- Scenario 3: Performance test with large data
```
Comment (ssihil) [inline sql/_13_issues/_24_1h/cases/cbrd_25164.sql]: Performance testing doesn't seem to have anything to do with this issue.
What is generate_series on line 78?? An error occurred during execution.

File: sql/_13_issues/_24_1h/cases/cbrd_25164.sql [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+create table tx (id int, va int);
+insert into tx (id, va) select generate_series, random()*100 from generate_series(1,100000);
+create index idx on tx(ln(va));
+update statistics on tx;
+select count(*) from tx where ln(va) > 4;
+
+drop table tx;
+
+-- Scenario 4: Changing the order in composite indexes
+-- Purpose: To verify the impact of changing the order of columns in composite indexes, including function columns, on query results and performance.
```
Comment (ssihil) [inline sql/_13_issues/_24_1h/cases/cbrd_25164.sql]:
```suggestion
-- Purpose: To verify the impact of changing the order of columns in composite indexes, including function columns, on query results
```

File: sql/_13_issues/_24_1h/cases/cbrd_25164.sql [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+create table tx (id int, va int, vb int, vc int);
+insert into tx values (1,1,10,100), (2,2,20,200), (3,3,30,300);
+
+-- Index creation with function
+create index idx on tx(va, ln(vb), vc);
+update statistics on tx;
+
+-- Query without index hint
+-- expected to use the function index created above, potentially leading to a covered index scan.
+select vc from tx where va = 1 and vc = 100;
```
Comment (ssihil) [inline sql/_13_issues/_24_1h/cases/cbrd_25164.sql]: It appears to be the same case as line 22.

File: sql/_13_issues/_24_1h/cases/cbrd_25164.sql [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+
+-- Query with index hint NO_COVERING_IDX
+-- explicitly instructs the optimizer not to use a covering index if available, which should force a different execution path, potentially affecting performance.
+select /*+ NO_COVERING_IDX */ vc from tx where va = 1 and vc = 100;
+
+-- Additional comparison for performance analysis
```
Comment (ssihil) [inline sql/_13_issues/_24_1h/cases/cbrd_25164.sql]: Performance testing doesn't seem to have anything to do with this issue.

File: sql/_13_issues/_24_1h/cases/cbrd_25164.sql [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+insert into tx values (1,10,100), (2,20,200);
+create index idx_va_ln on tx(va, ln(vb));
+create index idx_ln_va on tx(ln(vb), va);
+update statistics on tx;
+select id from tx where ln(vb) > 4 and va = 10;
+
+drop table tx;
+
+-- Scenario 5: Comparison with and without index hints
+-- Purpose: To compare the query results and performance with and without using index hints like NO_COVERING_IDX.
```
Comment (ssihil) [inline sql/_13_issues/_24_1h/cases/cbrd_25164.sql]:
```suggestion
-- Purpose: To compare the query results  with and without using index hints like NO_COVERING_IDX.
```

File: sql/_13_issues/_24_1h/cases/cbrd_25164.sql [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+
+drop index idx on tx;
+create index idx on tx(ln(va), vb, vc);
+update statistics on tx;
+select vb from tx where ln(va) > 0 and vc = 100;
+
+drop table tx;
+
+-- Scenario 1: Various function indexes and conditions in query
+-- Purpose: To verify the accuracy and performance of queries using function indexes with various mathematical functions such as ln(), exp(), sqrt().
```
Comment (ssihil) [inline sql/_13_issues/_24_1h/cases/cbrd_25164.sql]:
```suggestion
-- Purpose: To verify the accuracy of queries using function indexes with various mathematical functions such as ln(), exp(), sqrt().
```

File: sql/_13_issues/_24_1h/cases/cbrd_25164.sql [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+create index idx_sqrt on tx(sqrt(vc));
+update statistics on tx;
+select id from tx where ln(va) > 0;
+select id from tx where exp(vb) < 1000;
+select id from tx where sqrt(vc) > 30;
+
+drop table tx;
+
+-- Scenario 2: Queries including join operations
+-- Purpose: To verify the accuracy and performance of join operations on tables with function indexes.
```
Comment (ssihil) [inline sql/_13_issues/_24_1h/cases/cbrd_25164.sql]:
```suggestion
-- Purpose: To verify the accuracy of join operations on tables with function indexes.
```

File: sql/_13_issues/_24_1h/cases/cbrd_25164.sql [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+drop table if exists tx;
+drop table if exists ty;
+
+create table tx (id int, va int, vb int);
+create table ty (id int, tx_id int, vc int);
+insert into tx values (1,10,100), (2,20,200);
+insert into ty values (1,1,1000), (2,2,2000);
+create index idx on tx(ln(va));
+update statistics on tx;
+select ty.id, tx.va, ty.vc from tx join ty on tx.id = ty.tx_id where ln(tx.va) > 2;
```
Comment (ssihil) [inline sql/_13_issues/_24_1h/cases/cbrd_25164.sql]: Modified to use covered index.
create index idx on tx(id, ln(va), vb);
select ty.id, tx.vb, ty.vc from tx join ty on tx.id = ty.tx_id where vb > 2;

File: sql/_13_issues/_24_1h/cases/cbrd_25164.sql (hunk 4 above)
Comment (bagus-kim) [reply sql/_13_issues/_24_1h/cases/cbrd_25164.sql]: Please. Compare plan infomation.
Join graph terms is different.

--old version
csql> select /*+ NO_COVERING_IDX */ vc from tx where va = 1 and vc = 100;
Join graph segments (f indicates final):
seg[0]: [0]
seg[1]: vc[0] (f)
seg[2]: va[0]
Join graph nodes:
node[0]: dba.tx dba.tx(2/1) (sargs 0 1) (loc 0)
Join graph terms:
term[0]: [dba.tx].vc=100 (sel 0.001) (sarg term) (not-join eligible) (indexable vc[0]) (loc 0)
term[1]: [dba.tx].va=1 (sel 0.5) (sarg term) (not-join eligible) (indexable va[0]) (loc 0)

--patch version
csql> select /*+ NO_COVERING_IDX */ vc from tx where va = 1 and vc = 100;
Join graph segments (f indicates final):
seg[0]: [0]
seg[1]: vc[0] (f)
seg[2]: va[0]
Join graph nodes:
node[0]: dba.tx dba.tx(2/1) (sargs 0 1) (loc 0)
Join graph terms:
term[0]: [dba.tx].va=1 (sel 0.5) (sarg term) (not-join eligible) (indexable va[0]) (loc 0)
term[1]: [dba.tx].vc=100 (sel 0.001) (sarg term) (not-join eligible) (indexable vc[0]) (loc 0)

File: sql/_13_issues/_24_1h/cases/cbrd_25164.sql (hunk 1 above)
Comment (tw-kang) [reply sql/_13_issues/_24_1h/cases/cbrd_25164.sql]: review has been reflected. large data case has been removed.

File: sql/_13_issues/_24_1h/cases/cbrd_25164.sql (hunk 2 above)
Comment (tw-kang) [reply sql/_13_issues/_24_1h/cases/cbrd_25164.sql]: review has been reflected.

File: sql/_13_issues/_24_1h/cases/cbrd_25164.sql (hunk 5 above)
Comment (tw-kang) [reply sql/_13_issues/_24_1h/cases/cbrd_25164.sql]: review has been reflected.

File: sql/_13_issues/_24_1h/cases/cbrd_25164.sql (hunk 6 above)
Comment (tw-kang) [reply sql/_13_issues/_24_1h/cases/cbrd_25164.sql]: review has been reflected.

File: sql/_13_issues/_24_1h/cases/cbrd_25164.sql (hunk 7 above)
Comment (tw-kang) [reply sql/_13_issues/_24_1h/cases/cbrd_25164.sql]: review has been reflected.

File: sql/_13_issues/_24_1h/cases/cbrd_25164.sql (hunk 8 above)
Comment (tw-kang) [reply sql/_13_issues/_24_1h/cases/cbrd_25164.sql]: review has been reflected.

**Review by ssihil — APPROVED** [review]

**Review by ctshim — APPROVED** [review]

**Comment by ctshim** [comment]
I give you approval.
However, I have an additional comment.

The main point covered in [CBRD-25164] is not the plan, but fixing the issue of incorrect results.
I think it would be better to review the Plan separately in [CBRD-25256].

**Comment by tw-kang** [comment]
> I give you approval. However, I have an additional comment.
> 
> The main point covered in [CBRD-25164] is not the plan, but fixing the issue of incorrect results. I think it would be better to review the Plan separately in [CBRD-25256].

well noted. thank you for the review.

File: sql/_13_issues/_24_1h/cases/cbrd_25164.sql [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+-- The issue involves incorrect search results from covered index queries using function indexes.
+
+-- Verify correct results are returned for basic select query with function index.
+drop table if exists tx;
+
+create table tx (id int, va int, vb int, vc int, index idx(va, ln(va), vc));
+insert into tx values (1,1,1,1), (2,2,2,2);
+update statistics on tx;
+select vc from tx where va = 1;
+select vc from tx where va = 1 and vc > 0;
```
Comment (bagus-kim) [inline sql/_13_issues/_24_1h/cases/cbrd_25164.sql]: Add TestCase. Think about that.

select vc from tx where va = 1 using index idx(+);
select vc from tx where va = 1 using index tx.NONE;

File: sql/_13_issues/_24_1h/cases/cbrd_25164.sql (hunk 3 above)
Comment (tw-kang) [reply sql/_13_issues/_24_1h/cases/cbrd_25164.sql]: the duplicated query has been removed.
the review has been reflected.

File: sql/_13_issues/_24_1h/cases/cbrd_25164.sql (hunk 9 above)
Comment (tw-kang) [reply sql/_13_issues/_24_1h/cases/cbrd_25164.sql]: the review has been reflected.

**Review by bagus-kim — APPROVED** [review]

**Review by kwonhoil — APPROVED** [review]

**Pull request merged**